client-runtime/src/main/java/com/microsoft/rest/v3/RestProxy.java

```diff
@@ -406,94 +366,82 @@ private static Exception instantiateUnexpectedException(Class<? extends RestExce
+        // 'RestResponseBase' class instead.
+        Class<? extends RestResponse<?>> cls = (Class<? extends RestResponse<?>>) TypeUtil.getRawClass(entityType);
+        if (cls.equals(RestResponse.class)) {
+            cls = (Class<? extends RestResponse<?>>) (Object) RestResponseBase.class;
+        }
+
+        // we try to find the most specific constructor, which we do in the following order:
+        // 1) (HttpRequest httpRequest, int statusCode, HttpHeaders headers, Object body, Object deserializedHeaders)
+        // 2) (HttpRequest httpRequest, int statusCode, HttpHeaders headers, Object body)
+        // 3) (HttpRequest httpRequest, int statusCode, HttpHeaders headers)
+        List<Constructor<?>> ctors = Arrays.stream(cls.getDeclaredConstructors())
+                                             .filter(ctor -> {
+                                                 int paramCount = ctor.getParameterCount();
+                                                 return paramCount >= 3 && paramCount <= 5;
+                                             })
+                                             .sorted(Comparator.comparingInt(Constructor::getParameterCount))
+                                             .collect(Collectors.toList());
+
+        if (ctors.isEmpty()) {
+            throw new RuntimeException("Cannot find suitable constructor for class " + cls);
+        }
+
+        // try to create an instance using our list of potential candidates
+        for (int i = 0; i < ctors.size(); i++) {
+            Constructor<? extends RestResponse<?>> ctor = (Constructor<? extends RestResponse<?>>) ctors.get(i);
+
+            try {
+                final int paramCount = ctor.getParameterCount();
+
+                switch (paramCount) {
+                    case 3: {
+                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders);
+                    }
+                    case 4: {
+                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject);
+                    }
+                    case 5: {
+                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject, response.decodedHeaders().block());
+                    }
                 }
-            } else {
-                // For now we're just throwing if the Maybe didn't emit a value.
-                asyncResult = handleBodyReturnType(response, methodParser, entityType);
+            } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
+                throw reactor.core.Exceptions.propagate(e);
             }
-
-            return asyncResult;
-        } catch (ReflectiveOperationException e) {
-            throw new RuntimeException(e);
         }
+
+        // error
+        throw new RuntimeException("Cannot find suitable constructor for class " + cls);
```

Comment:
Is it possible to add this as the `default:` case? I think it makes it easier to follow rather than falling out of the switch case. IIRC, spotbugs has a check where it would give you a warning if there was no default case, too.

Reply:
Never mind. I see that its in a for loop. :) Ignore this.

Reply:
I considered that (because this is my preference too), however, there should never be a default case - we've filtered to only get constructors with between 3 and 5 params.